### PR TITLE
6L/512d depth scale-up (vs current 4L/512d baseline)

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

The current yi architecture is 4L/512d/8h (4 layers, 512 hidden dim, 8 heads). This was a tuned win over 2L and 3L configurations. Going deeper to **6 layers** at the same 512d width adds ~50% more model capacity with modest compute increase. Deeper models can learn more complex feature hierarchies — important for the τ_y/τ_z gap which requires the model to understand both local surface geometry and global flow context simultaneously.

The hypothesis: 6L/512d will learn better representations for complex wall shear patterns, closing the τ_y/τ_z gap beyond what 4L achieves.

## Instructions

Run the yi baseline config with increased depth:

**Arm A: 6L/512d/8h**
```bash
cd target/
CUDA_VISIBLE_DEVICES=0,1,2,3 torchrun --nproc_per_node=4 train.py \
  --agent norman --wandb-group yi-round36-depth-scaleup \
  --learnable-pe --optimizer lion --lr 1e-4 --weight-decay 5e-4 --clip-grad-norm 0.5 \
  --grad-ema-alpha 0.5 \
  --surface-curvature-features k1_k2 --beta-nll-beta 0.5 \
  --model-layers 6 --model-hidden-dim 512 --model-heads 8 --model-slices 128 \
  --batch-size 4 --validation-every 1 --no-compile-model \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536
```

**Arm B: 8L/512d/8h** (if Arm A fits in GPU memory and shows promise at epoch 5)
```bash
cd target/
CUDA_VISIBLE_DEVICES=0,1,2,3 torchrun --nproc_per_node=4 train.py \
  --agent norman --wandb-group yi-round36-depth-scaleup \
  --learnable-pe --optimizer lion --lr 1e-4 --weight-decay 5e-4 --clip-grad-norm 0.5 \
  --grad-ema-alpha 0.5 \
  --surface-curvature-features k1_k2 --beta-nll-beta 0.5 \
  --model-layers 8 --model-hidden-dim 512 --model-heads 8 --model-slices 128 \
  --batch-size 4 --validation-every 1 --no-compile-model \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536
```

**Memory check**: First, verify that 6L/512d fits in memory with batch-size=4. If OOM, reduce batch-size to 2. If still OOM, reduce model-slices to 64.

**Decision at epoch 5:**
- Val_abupt < 8.5%: continue Arm A to completion, launch Arm B.
- Val_abupt 8.5–9.0%: continue Arm A only, skip Arm B.
- Val_abupt > 9.5%: stop, report (model may be over-capacity for data size).

Report τ_y, τ_z, τ_x, surface_pressure, volume_pressure, val_abupt per epoch. Include W&B run IDs. Also note GPU memory usage.

## Baseline (yi, PR #576)

| Metric | yi best | AB-UPT target | Gap |
|--------|---------|---------------|-----|
| val_abupt | **8.2528%** | — | — |
| surface_pressure | 5.127% | 3.82% | 1.34× |
| wall_shear | 9.233% | 7.29% | 1.27× |
| volume_pressure | 12.438% | 6.08% | **2.05×** |
| τ_x | 7.815% | 5.35% | 1.46× |
| τ_y | 9.233% | 3.65% | **2.53×** |
| τ_z | 10.449% | 3.63% | **2.88×** |

Beat `val_abupt < 8.2528%` to land on yi.

### CODE REVIEW

**Comment by morganmcg1** [comment]
STUDENT norman: Arm A (6L/512d) is running on GPUs 0-3 (W&B run `pf7qeig9`, group `yi-round36-depth-scaleup`). Started 20:12 UTC, currently at step 714/5442 (~13%) of epoch 1, 1.84 s/it.

**Timing reality check:** at this pace each epoch takes ~167 min. The train timeout is `SENPAI_TIMEOUT_MINUTES (360) - SENPAI_VAL_BUDGET_MINUTES (90) = 270 min`, so we'll get **~1.6 epochs** of training before the run hard-stops and runs final val/test (~90 min eval). Reaching the PR's epoch-5 decision point isn't possible inside one launch — I'll grade Arm A on its first validation (end of epoch 1, ~22:59 UTC) plus the mid-epoch timeout validation in epoch 2.

Given Arm A alone consumes the full timeout, **launching Arm B (8L/512d) in parallel is not feasible** with my 4 GPUs. If Arm A clearly underperforms 4L, I'll abort early and report. If it tracks or beats 4L baseline at epoch 1, I'll let it continue to timeout.

Will report results once the first validation lands.

**Comment by morganmcg1** [comment]
ADVISOR: Good situational awareness on the budget — your 1.6-epoch read is correct and the right way to think about this. Here are your revised gates:

**Revised decision gates for 6L/512d (from-scratch, single epoch budget):**
- EP1 val_abupt ≤ 12%: continue to run timeout (this is from-scratch, 4L EP1 baseline was ~10-11%)
- EP1 val_abupt 12–15%: marginal — look at the trend slope; if loss is still visibly dropping, let it run
- EP1 val_abupt > 15%: abort and report — too far behind for depth to help within budget

**Why EP1 ≤ 12% instead of ≤ 10%:** From-scratch 6L models train slower initially due to more parameters and gradient flow depth. The decisive question is whether the 6L model converges to something better than the 4L model at the end of its available training time, not whether it matches 4L at epoch 1.

**Primary signal:** End-of-budget val_abupt vs the yi merge bar of **8.2528%**. If you land below this on the EMA model, this is a merge candidate.

**Arm B (8L):** Since you confirmed 4 GPUs only, skip Arm B. The sequential cost is too high — if 6L wins, we can run 8L as a separate PR.

**One critical note:** Your Arm A command includes --grad-ema-alpha 0.5 and --surface-curvature-features k1_k2 and --beta-nll-beta 0.5 — these are NOT in the yi baseline (PR #576). Please confirm you intend to test 6L depth as the isolated variable, or if you have intentionally layered these on top. If it is a layered run, report each metric so we can factor out the depth contribution.

Report: EP1 val_abupt + τ_y + τ_z + vol_p + step timing. Update when EP1 lands (~22:59 UTC).

**Comment by morganmcg1** [comment]
STUDENT norman: Picking this up now.

**Re: layered features.** Yes, this is intentionally layered, not a clean depth-isolation experiment. The flags in your PR command match the active yi merge bar plus two extras:

- `--learnable-pe`, `--optimizer lion`, `--lr 1e-4`, `--weight-decay 5e-4`, `--clip-grad-norm 0.5`, `--grad-ema-alpha 0.5` — these are the PR #576 yi SOTA stack (val 8.2528%).
- `--surface-curvature-features k1_k2` — PR #580 (haku) infra feature, NOT on the merge bar.
- `--beta-nll-beta 0.5` — PR #583 (edward) loss feature, NOT on the merge bar.

So the run actually tests **6L depth + (k1_k2 curvature + β-NLL) layered on the yi PR #576 stack**. If 6L wins, depth alone vs k1_k2/β-NLL contribution will need a separate ablation. I'll quote all primary metrics so we can compare against PR #576 directly and against the AB-UPT τ_y/τ_z targets.

**Run status (21:25 UTC):** EP1 step 2385/5442 (~44%), 1.84 s/it, 1:33 ETA → first val ~22:58 UTC. W&B run `pf7qeig9` is healthy (no NaN/Inf, grad-norm tracking clip threshold). 2437 logged steps. GPU mem 96/97 GB across all 4 — about 1.4 GB headroom, no OOM risk.

Will report at EP1 val and then again at training timeout / EP2 val.

**Comment by morganmcg1** [comment]
STUDENT norman:
SENPAI-RESULT: {"terminal":true,"status":"complete","pending_arms":false,"wandb_run_ids":["pf7qeig9"],"primary_metric":{"name":"val_primary/abupt_axis_mean_rel_l2_pct","value":18.8959},"test_metric":{"name":"val_primary/abupt_axis_mean_rel_l2_pct","value":18.8959}}

## Results — Arm A (6L/512d, layered features) ABORTED at EP1

Per the revised gate **EP1 val_abupt > 15% → abort and report**, I killed the run after EP1 validation landed at **18.8959%**. No EP2 was run.

### EP1 validation (W&B run `pf7qeig9`, step 5442)

| Metric | EP1 (6L+layered) | yi PR #576 best (4L) | AB-UPT target |
|---|---|---|---|
| **val_abupt** | **18.8959%** | 8.2528% | — |
| surface_pressure | 11.8498% | 5.127% | 3.82% |
| wall_shear | 18.9183% | 9.233% | 7.29% |
| τ_x | 15.8048% | 7.815% | 5.35% |
| τ_y | 22.8913% | 9.233% | 3.65% |
| τ_z | 25.0254% | 10.449% | 3.63% |
| volume_pressure | 18.9083% | 12.438% | 6.08% |

### Run details
- **Train EP1**: 167.6 min (10057s), 5442 steps @ 1.85 s/it, train_loss=−0.150
- **Validation**: ~14 min (EP1 train ended ~22:59 UTC, val landed 23:13 UTC)
- **Peak GPU memory**: 98.4 GB (4× H100, ~1.4 GB headroom)
- **Model**: 18.52M params (vs 12.70M for 4L/512d at PR #576)
- **W&B**: https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml/runs/pf7qeig9

### Command
```bash
CUDA_VISIBLE_DEVICES=0,1,2,3 torchrun --standalone --nproc_per_node=4 train.py \
  --agent norman --wandb-group yi-round36-depth-scaleup --wandb-name norman/6L-512d-armA \
  --learnable-pe --optimizer lion --lr 1e-4 --weight-decay 5e-4 --clip-grad-norm 0.5 \
  --grad-ema-alpha 0.5 \
  --surface-curvature-features k1_k2 --beta-nll-beta 0.5 \
  --model-layers 6 --model-hidden-dim 512 --model-heads 8 --model-slices 128 \
  --batch-size 4 --validation-every 1 --no-compile-model \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536
```

### Train-loss trajectory (sanity check the model was actually learning)
| step | train/loss | tau_x | tau_y | tau_z |
|---|---|---|---|---|
| 0 | 1.219 | 1.378 | 1.488 | 1.224 |
| 1000 | -0.082 | 0.266 | 0.980 | 0.535 |
| 3000 | -0.314 | 0.050 | 0.088 | 0.102 |
| 5000 | -0.292 | 0.022 | 0.033 | 0.041 |
| 5331 | -0.295 | 0.019 | 0.029 | 0.033 |

Train loss is dropping smoothly, no NaN/Inf, gradients clipped to 0.5 (pre-clip norms ~1.2). The model is learning — it's just **far behind 4L+yi at the same checkpoint**.

### What happened — analysis

**The run cleared the abort gate by a wide margin (18.9% vs 15% gate).** Three observations:

1. **6L is starting from much further behind than 4L.** From-scratch 4L at the same point (EP1) was reportedly ~10–11% per the advisor. 6L is at 18.9% — that's a 7–9 pp gap, roughly 2× worse. Even granting that 6L converges slower at the start, closing this gap inside the remaining ~85 min of training (~50% of EP2) is implausible.

2. **τ_y / τ_z (the components the hypothesis was supposed to address) are the worst.** τ_z = 25.0%, τ_y = 22.9%. These are 2.5× higher than the 4L baseline. The hypothesis was that depth would help τ_y/τ_z by combining local geometry and global flow context — at this checkpoint we have the opposite: deeper model, but worse τ_y/τ_z than shallower.

3. **The layered features may be a confound, not just a co-treatment.** This was layered (6L + k1_k2 + β-NLL) on top of yi. β-NLL fundamentally changes the loss surface (variance head, negative-loss regime). Stacking deeper architecture + new loss + new geometric features all at once is a known recipe for slow convergence: each piece needs different hyperparameters, and 4L's tuned LR/wd may simply not be right for any of the three new variables.

### Suggested follow-ups

1. **Clean depth ablation.** Re-run 6L/512d with the **plain yi PR #576 stack** (no k1_k2, no β-NLL). This isolates depth as the variable. If clean 6L is comparable or better than 4L on EP1, depth is the right direction and the layered features were the problem.
2. **6L with longer LR warmup or higher LR.** Deeper Lion-trained models often need either longer warmup (currently 1 epoch = 5442 steps) or higher peak LR for the deeper layers to start contributing. With 4 GPUs and 1.6 epoch budget, this is the wrong setup to discover that.
3. **Skip 8L for now.** 6L+layered already burned full budget on a single arm; 8L would be even slower per epoch.
4. **Sequential layered tests.** If we want to combine depth + k1_k2 + β-NLL, do them one at a time on a confirmed 4L baseline first, then stack onto whichever depth wins.

### Final terminal status
- Aborted at EP1 per advisor gate. No checkpoint saved beyond EP1 best.
- 4 GPUs (0–3) freed, no other runs queued.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #633 because ADVISOR: Closing — 6L/512d depth scale-up failed the EP1 abort gate (18.9959% >> 15% threshold, vs yi merge bar 8.2528%).

**Why this failed:** 6L from scratch at the 4L-tuned LR (1e-4) trains much slower than 4L. At EP1, 6L is at 18.9% while 4L reaches ~10-11% at the same checkpoint — roughly a 7-9 pp gap. There's no headroom to close this inside the remaining budget.

**The confound:** This run also layered k1_k2 curvature features + β-NLL loss on top of depth, making it impossible to attribute any future improvement (or failure) to depth alone. The student's analysis is correct: multiple simultaneous changes with a non-standard loss surface (β-NLL) likely hurt convergence.

**Conclusion on depth scaling:** 6L depth at the current 4L-tuned hyperparameters is not viable at our budget. If the team wants to revisit depth, it would need: (a) a clean isolated depth ablation without k1_k2/β-NLL, (b) longer LR warmup or slightly higher LR to allow deeper layers to activate properly, and (c) ideally 3+ epochs of budget. This remains a valid research direction for a higher-compute setting.

**Norman's suggested follow-ups are good** — especially the clean depth ablation idea. That will be the basis for the next assignment.

Recording as: **CLOSED — depth scaling null at current budget (6L EP1 18.9% >> 4L 8.25%).**.
